tests/test_peft.py

```diff
@@ -0,0 +1,115 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tests related to PEFT integration."""
+
+import json
+from pathlib import Path
+
+import pytest
+import torch
+from peft import AutoPeftModelForCausalLM, LoraConfig, PeftModel
+from peft import get_peft_model as orig_get_peft_model
+from safetensors.torch import load_file
+from transformers import LlamaForCausalLM
+
+from optimum.neuron import get_peft_model
+from optimum.neuron.utils.peft_utils import NeuronPeftModel
+from optimum.neuron.utils.testing_utils import is_trainium_test
+
+from . import DistributedTest
+from .utils import create_accelerator, create_static_seed_patcher, get_tokenizer_and_tiny_llama_model
+
+
+def get_peft_config():
+    target_modules = ["q_proj", "k_proj", "v_proj", "o_proj", "gate_proj", "up_proj", "down_proj"]
+    return LoraConfig(
+        r=4, lora_alpha=16, target_modules=target_modules, lora_dropout=0.1, bias="none", task_type="CAUSAL_LM"
+    )
+
+
+def test_get_peft_model():
+    peft_config = get_peft_config()
+    _, model = get_tokenizer_and_tiny_llama_model()
+    orig_peft_model = orig_get_peft_model(model, peft_config)
+
+    assert isinstance(orig_peft_model, PeftModel)
+    assert not isinstance(orig_peft_model, NeuronPeftModel)
+
+    _, model = get_tokenizer_and_tiny_llama_model()
+    peft_model = get_peft_model(model, peft_config)
+
+    assert isinstance(peft_model, NeuronPeftModel)
+
+
+@is_trainium_test
+class TestPeft(DistributedTest):
+    @pytest.fixture(
+        scope="class",
+        params=[[2, 1, 1]],
+        ids=["dp=2"],
+    )
+    def parallel_sizes(self, request):
+        return request.param
+
+    @pytest.mark.world_size(2)
+    def test_peft_model_is_converted_to_neuron_peft_model(self):
+        model = AutoPeftModelForCausalLM.from_pretrained("peft-internal-testing/tiny-random-BertModel-lora")
+        assert isinstance(model, PeftModel)
+        accelerator = create_accelerator(1, 1)
+        model = accelerator.prepare(model)
+        assert isinstance(model, NeuronPeftModel)
+
+    def test_save_pretrained(self, parallel_sizes, tmpdir):
+        _, tp_size, pp_size = parallel_sizes
+
+        output_dir = Path(tmpdir)
+
+        peft_config = get_peft_config()
+
+        # PEFT model saved using `PeftModel`.
+        seed_patcher = create_static_seed_patcher(LlamaForCausalLM, 42)
+        with seed_patcher:
+            _, model = get_tokenizer_and_tiny_llama_model()
+            orig_model_path = output_dir / "orig_peft"
+            orig_peft_model = orig_get_peft_model(model, peft_config)
+
+        orig_peft_model.save_pretrained(orig_model_path.as_posix())
+
+        # PEFT model saved using `NeuronPeftModel`.
+        seed_patcher = create_static_seed_patcher(LlamaForCausalLM, 42)
```

Comment:
Do you really need to recreate this ?

Reply:
Yes, otherwise it fails (tried that). Maybe we can make it "reusable" because it's annoying but did not want to spend too much time on that.